orttraining/orttraining/core/optimizer/qdq_fusion.cc

```diff
@@ -0,0 +1,129 @@
+  // Replace the original zero point initializer with the float zero point initializer
+
+  auto zero_point_tensor_int = graph.GetInitializer(zero_point_initializer_name, true);
+  ONNX_NAMESPACE::TensorProto zero_point_tensor_float;
+  zero_point_tensor_float.set_name(graph.GenerateNodeArgName(zero_point_tensor_int->name()));
+  zero_point_tensor_float.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+  for (const auto val : zero_point_tensor_int->int32_data()) {
+    zero_point_tensor_float.add_float_data(static_cast<float>(val));
+  }
+  for (const auto& dim : zero_point_tensor_int->dims()) {
+    zero_point_tensor_float.add_dims(dim);
+  }
+  graph.RemoveInitializedTensor(zero_point_tensor_int->name());
+  return graph_utils::AddInitializer(graph, zero_point_tensor_float);
+}
+
+void FuseQDQNodes(Graph& graph, Node& quantize_node, Node& dequantize_node, NodeArg& zero_point_node_arg) {
+  const InlinedVector<NodeArg*> fake_quant_inputs{
+      quantize_node.MutableInputDefs()[0], quantize_node.MutableInputDefs()[1], &zero_point_node_arg};
+  ONNX_NAMESPACE::TypeProto bool_mask_tensor;
+  bool_mask_tensor.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_BOOL);
+  NodeArg& fake_quant_gradient_mask =
+      graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("fake_quant_gradient_mask"), &bool_mask_tensor);
+  const InlinedVector<NodeArg*> fake_quant_outputs{dequantize_node.MutableOutputDefs()[0],
+                                                   &fake_quant_gradient_mask};
+  Node& fake_quant_node = graph.AddNode(graph.GenerateNodeName("FakeQuant"), "FakeQuant",
+                                        "FakeQuant after fusion of " + quantize_node.Name() + " and " +
+                                            dequantize_node.Name(),
+                                        fake_quant_inputs, fake_quant_outputs, {}, kMSDomain);
+
+  // The quant_min and quant_max attribute values are set based on the elem_type of the zero point tensor.
+  // If the type is INT8, the quant_min and quant_max are set to -128 and 127 respectively. Else UINT8 is
+  // assumed, and quant_min and quant_max are set to 0 and 255 respectively.
+  ORT_ENFORCE(quantize_node.InputDefs()[2]->TypeAsProto(), "QuantizeLinear node ", quantize_node.Name(),
+              " zero_point type proto is not set.");
+  const auto zero_point_type = quantize_node.InputDefs()[2]->TypeAsProto()->tensor_type().elem_type();
+  fake_quant_node.AddAttribute(
+      "quant_min", static_cast<int64_t>(zero_point_type == ONNX_NAMESPACE::TensorProto_DataType_INT8 ? -128 : 0));
+  fake_quant_node.AddAttribute(
+      "quant_max", static_cast<int64_t>(zero_point_type == ONNX_NAMESPACE::TensorProto_DataType_INT8 ? 127 : 255));
+
+  // Set execution provider for FakeQuant
+  fake_quant_node.SetExecutionProviderType(quantize_node.GetExecutionProviderType());
+
+  // Clean up post fusion
```

Comment:
Replaced current logic so that we can use `FinalizeNodeFusion`.

The way I do it now is replace the original quantize node's zero point input to be the new initializer (float). And then just simply call `FinalizeNodeFusion`. Following this, I add the extra output.